{
  try {
    while (true) {
      Exception e=null;
      try {
        if (fn == null) {
          LOG.info("fn==null," + fn.getClass());
          continue;
        }
        fn.run();
        e=fn.error();
      }
 catch (      Exception ex) {
        e=ex;
      }
      if (e != null) {
        throw e;
      }
      Object rtn=fn.getResult();
      if (this.needSleep(rtn)) {
        return;
      }
    }
  }
 catch (  InterruptedException e) {
    LOG.info("Async loop interrupted!");
  }
catch (  Exception e) {
    if (e.getCause() instanceof InterruptedException) {
      LOG.info("Async loop interrupted!");
    }
 else {
      LOG.error("Async loop died!",e);
      killfn.execute(e);
    }
  }
}
